ignition-server/controllers/machineconfigserver_ignitionprovider.go

```diff
@@ -0,0 +1,338 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/openshift/hypershift/control-plane-operator/releaseinfo"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	k8sutilspointer "k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ IgnitionProvider = (*MCSIgnitionProvider)(nil)
+
+// MCSIgnitionProvider is an IgnitionProvider that uses
+// MachineConfigServer pods to build ignition payload contents.
+type MCSIgnitionProvider struct {
+	Client          client.Client
+	ReleaseProvider releaseinfo.Provider
+	Namespace       string
+}
+
+func (p *MCSIgnitionProvider) GetPayload(ctx context.Context, tokenSecret *corev1.Secret) (payload []byte, err error) {
+	// TODO(alberto): If the MCS supports binding address
+	// https://github.com/openshift/machine-config-operator/pull/2630/files
+	// we could bind it to localhost and get the payload by execing
+	// https://zhimin-wen.medium.com/programing-exec-into-a-pod-5f2a70bd93bb
+	// Otherwise with current approach the mcs pod is temporary exposed in the pod network every time a payload is generated.
+	releaseImage := string(tokenSecret.Data[TokenSecretReleaseKey])
+	img, err := p.ReleaseProvider.Lookup(ctx, releaseImage)
+	if err != nil {
+		return nil, fmt.Errorf("failed to look up release image metadata: %w", err)
+	}
+
+	mcsResourceName := tokenSecret.GetName()
+	mcsServiceAccount := MachineConfigServerServiceAccount(p.Namespace, mcsResourceName)
+	mcsRoleBinding := MachineConfigServerRoleBinding(p.Namespace, mcsServiceAccount)
```

Comment:
I didn't realize we'd have to install a service account and role binding for these pods. The way this is handled here isn't thread-safe. Although the controller is single threaded, still seems like dangerous territory. If we absolutely have to do this, what if we made an immutable unique SA and rolebinding for the token request? Is there any way around the SA/rolebinding requirement?

Reply:
>what if we made an immutable unique SA and rolebinding for the token request

do you mean one for all of them? I can try that

Reply:
Using one for all of them makes this code feel more like a controller actively managing resources which feels like a step backwards to me... I was suggesting that each distinct token secret would have its own SA and role binding to drive the payload lookup

Reply:
this is the case already